src/main/java/ticTacToe/player/SmartComputer.java

```diff
@@ -0,0 +1,94 @@
+package ticTacToe.player;
+
+import ticTacToe.game.Lines;
+import ticTacToe.grid.Grid;
+import ticTacToe.ui.Ui;
+
+import java.util.*;
+
+import static ticTacToe.game.Mark.switchPlayerMark;
+
+public class SmartComputer implements Player {
+
+    private final String mark;
+    private final MoveGenerator firstMove;
+    private int score;
+
+    public SmartComputer(String mark, MoveGenerator firstMove) {
+        this.mark = mark;
+        this.firstMove = firstMove;
+    }
+
+    public String makeMove(Ui ui, Grid grid, Lines lines) {
+        StringBuilder move = new StringBuilder();
+        if (grid.isAllEmpty()) {
+            move.append(firstMove.getRandomMove(grid));
+        } else {
+            move.append(bestMovePosition(grid, lines));
+        }
+        ui.confirmMove(mark, move.toString());
```

Comment:
I'm a bit surprised that the computer player needs a Ui. And I saw that you're repeating this method call `ui.confirmMove` in the HumanPlayer as well. Maybe you could call this method in GameFlow? Then you don't need to repeat it.

Reply:
I liked the fact that in GameFlow the method gameFlow() doesn't contain any calls of Ui methods but only the actual 'logic' of the game... but that doesn't mean that I'm doing the right thing :)

Reply:
I was thinking that you passed the Ui into GameFlow anyway, so you've got it available in the whole class and can use it.